### PR TITLE
completed Lab1

### PR DESCRIPTION
**Pull request by mewaeltsegay**
Created a folder inside the students directory.

### CODE REVIEW

**Comment by mewaeltsegay** [comment]
Completed Lab1

File: tasks/task-01/README.md
Comment (ADmangarakov) [inline tasks/task-01/README.md:1]: Do not change the task

File: students/mt-desta/lab2/source/__pycache__/knn.cpython-312.pyc
Comment (ADmangarakov) [inline students/mt-desta/lab2/source/__pycache__/knn.cpython-312.pyc:1]: Do not commit pycache

File: students/mt-desta/lab1/README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#### 2. I have implented a hierarchical algorithm based on Agglomerative clustering. [my_hierarchy.py](./source/my_hierarchy.py) contains the code for this implementation. It takes the number of desired clusters, linkage method and a dataframe as arguements. It outputs the linkage matrix and labels. I have implemented a function tha takes these and plots a dendrogram.
+
+<img src="img/image3.png">
+<img src="img/image4.png">
+
+#### 2.1 Dendrograms for the two datasets
+
+<img src="img/image5.png">
+        
+#### 2.2 Optimum number of Clusters
```
Comment (ADmangarakov) [inline students/mt-desta/lab1/README.md:24]: Calculate the optimal number of clusters based on the maximum increment R_t for hierarchical clustering based on Ward distance

File: students/mt-desta/lab1/source/em.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            diff = X - self.means[k].reshape(1,-1)
+            maha_dist = np.sum(np.dot(diff, np.linalg.inv(self.covariances[k])) * diff, axis=1)
+            likelihoods[:, k] = self.weights[k] * np.exp(-0.5 * maha_dist) / np.sqrt((2 * np.pi) ** X.shape[1] * np.linalg.det(self.covariances[k]))
+        
+        # Assign each sample to the cluster with the highest likelihood
+        predictions = np.argmax(likelihoods, axis=1)
+        
+        return predictions
+    
+# df = pd.read_csv('data/Mall_Customers.csv')[['Annual Income (k$)', 'Spending Score (1-100)']]
```
Comment (ADmangarakov) [inline students/mt-desta/lab1/source/em.py:66]: Remove comments

File: students/mt-desta/lab1/README.md (hunk 1 above)
Comment (mewaeltsegay) [reply students/mt-desta/lab1/README.md:31]: resolved issue.

File: students/mt-desta/lab1/README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+####  <p style='text-aligh: justify'> 1.3 The first dataset is made up of **clusters with jumpers**. As clearly seen on the plot we have one or two points that connect the areas with dense consentration of points. As for the second dataset, it has no clear shape but I assume is made up of **overlapping clusters**.  </p>
+
+#### <p style='text-aligh: justify'> 1.4 Though it is not clear, for the first dataset my assumption is at least 5 clusters by looking at the graph. As for the second dataset, there seem to be 2 or 3 areas with consentration but I cannot be so sure.</p>
+
+#### 2. I have implented a hierarchical algorithm based on Agglomerative clustering. [my_hierarchy.py](./source/my_hierarchy.py) contains the code for this implementation. It takes the number of desired clusters, linkage method and a dataframe as arguements. It outputs the linkage matrix and labels. I have implemented a function tha takes these and plots a dendrogram.
+
+<img src="img/image3.png">
+<img src="img/image4.png">
+
+#### 2.1 Dendrograms for the two datasets
```
[the remark this answers is not in the document]
Comment (mewaeltsegay) [reply students/mt-desta/lab1/README.md:20]: done.

File: students/mt-desta/lab1/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+####    * measured the clustering speed;
+
+### For Mall Customers: 
+
+| | Metric |	Implemented Algorithms |	Library Algorithm |
+| - |----- | ----------------------- | ------------------ |
+|1 |	Average Inter-Cluster Distance (Agglomerative) |	59.040257 |	58.271701 |
+| 2 |	Average Inter-Cluster Distance (EM) |	52.191699 |	56.166705 |
+| 3 |	Average Inter-Cluster Distance (DBSCAN) |	53.652448 |	46.076062 |
+| 4 |	Average Intra-Cluster Distance (Agglomerative) |	304.787110 |	24.506988 |
```
Comment (ADmangarakov) [inline students/mt-desta/lab1/README.md:68]: Why are the values ​​so different?

File: students/mt-desta/lab1/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+
+### For IRIS Dataset: 
+
+| | Metric |	Implemented Algorithms |	Library Algorithm |
+| - |----- | ----------------------- | ------------------ |
+|0|	Average Inter-Cluster Distance (Agglomerative)|	1.531983|	2.745475|
+|1|	Average Inter-Cluster Distance (EM)|	2.740543|	0.586880|
+|2|	Average Inter-Cluster Distance (DBSCAN)|	2.757433|	2.300989|
+|3|	Average Intra-Cluster Distance (Agglomerative)|	5.510851|	0.530974|
```
Comment (ADmangarakov) [inline students/mt-desta/lab1/README.md:83]: Again, why implemented algorithm is ten times better?

File: students/mt-desta/lab1/README.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+#### 5. The following metrics were applied on reference algorithms.
+####    * Based on the clustering results, calculated the average intra-cluster distance;
+####    * Based on the clustering results, calculated the average inter cluster distance;
+####    * measured the clustering speed;
+
+#### 6. Comparisons between implemented and reference algorithms is the following.
+
+<img src="img/image10.png">
+
+#### <p> Average Inter-Cluster Distance: Lower values indicate better separation between clusters.
```
Comment (ADmangarakov) [inline students/mt-desta/lab1/README.md:99]: Why so?

File: students/mt-desta/lab1/README.md (hunk 4 above)
Comment (mewaeltsegay) [reply students/mt-desta/lab1/README.md]: I thought I wrote better code but it turns out there was a bug in the code. The library code returns cluster labels not distances, so my calculations were wrong. Resolved the issues.

**Comment by mewaeltsegay** [comment]
Removed a bug that made it seem the implemented hierarchical algorithm was performing better. Now the correct results are shown.

**Pull request closed without merging**